pennylane/interfaces/batch/jax.py

```diff
@@ -1,299 +1,278 @@
-
-            return (tuple(res),)
-
-        # Gradient function is a device method.
-        with qml.tape.Unwrap(*tapes):
-            jacs = gradient_fn(tapes, **gradient_kwargs)
-
-        vjps = [qml.gradients.compute_vjp(d, jac) for d, jac in zip(g, jacs)]
-        res = [[jnp.array(p) for p in v] for v in vjps]
-        return (tuple(res),)
-
-    wrapped_exec.defvjp(wrapped_exec_fwd, wrapped_exec_bwd)
-    return wrapped_exec(params)
-
-
-# The execute function in forward mode
-def _execute_with_fwd(
-    params,
-    tapes=None,
-    device=None,
-    execute_fn=None,
-    gradient_kwargs=None,
-    _n=1,
-):  # pylint: disable=dangerous-default-value,unused-argument
-
-    _validate_tapes(tapes)
-
-    # Only have scalar outputs
-    total_size = len(tapes)
-
-    @jax.custom_vjp
-    def wrapped_exec(params):
-        def wrapper(p):
-            """Compute the forward pass by returning the jacobian too."""
-            new_tapes = []
-
-            for t, a in zip(tapes, p):
-                new_tapes.append(t.copy(copy_operations=True))
-                new_tapes[-1].set_parameters(a)
-
-            res, jacs = execute_fn(new_tapes, **gradient_kwargs)
-
-            # On the forward execution return the jacobian too
-            return res, jacs
-
-        fwd_shapes = [jax.ShapeDtypeStruct((1,), dtype) for _ in range(total_size)]
-        jacobian_shape = [jax.ShapeDtypeStruct((1, len(p)), dtype) for p in params]
-        res, jacs = host_callback.call(
-            wrapper,
-            params,
-            result_shape=tuple([fwd_shapes, jacobian_shape]),
-        )
-        return res, jacs
-
-    def wrapped_exec_fwd(params):
-        res, jacs = wrapped_exec(params)
-        return res, tuple([jacs, params])
-
-    def wrapped_exec_bwd(params, g):
-
-        # Use the jacobian that was computed on the forward pass
-        jacs, params = params
-
-        # Adjust the structure of how the jacobian is returned to match the
-        # non-forward mode cases
-        # E.g.,
-        # [DeviceArray([[ 0.06695931,  0.01383095, -0.46500877]], dtype=float32)]
-        # is mapped to
-        # [[DeviceArray(0.06695931, dtype=float32), DeviceArray(0.01383095,
-        # dtype=float32), DeviceArray(-0.46500877, dtype=float32)]]
-        res_jacs = []
-        for j in jacs:
-            this_j = []
-            for i in range(j.shape[1]):
-                this_j.append(j[0, i])
-            res_jacs.append(this_j)
-        return tuple([tuple(res_jacs)])
-
-    wrapped_exec.defvjp(wrapped_exec_fwd, wrapped_exec_bwd)
-    res = wrapped_exec(params)
-
-    tracing = any(isinstance(r, jax.interpreters.ad.JVPTracer) for r in res)
-
-    # When there are no tracers (not differentiating), we have the result of
-    # the forward pass and the jacobian, but only need the result of the
-    # forward pass
-    if len(res) == 2 and not tracing:
-        res = res[0]
-
-    return res
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
```

Comment:
This file was renamed locally using `git mv`, hence GitHub shows major changes here. Leaving a comment at spots where a change occurred in reality.